File: Python/problem046.py
#! /usr/bin/env python3
'''
Problem 46 - Project Euler
http://projecteuler.net/index.php?section=problems&id=046
'''
import math
from mathtools import primes

if __name__ == '__main__':
    N = 1_000_0
    ps = primes(N)
    odd_composite_num = list(set(range(5, N+1, 2)) - set(ps))
    for i in odd_composite_num:
        for sq in range(1, int(math.sqrt(i/2))+1)[::-1]:
            if (i - 2 * (sq ** 2)) in ps:
                break
        else:
            print(i)
            break

File: Clojure/problem046.clj
#! /usr/bin/env clojure
; Problem 46 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=46
(require '[mathtools :refer [primes]])

(let [N 10000
      ps (primes N)
      odd-composite-num (->> (range 5 (inc N) 2)
                             (filter #(< (.indexOf ps %) 0)))]
  (loop [odds odd-composite-num]
    (let [odd (first odds)
          g (for [sq (reverse (range 1 (inc (int (Math/sqrt (/ odd 2))))))
                      :let [i (- odd (* 2 (* sq sq)))]
                      :when (>= (.indexOf ps i) 0)] i)]
      (if (empty? g)
        (println odd)
        (recur (rest odds))))))
